File: backend/src/api/v1/qualifications.py
"""Qualification upload and review endpoints (US2).

Promoter-facing endpoints for uploading files and managing qualification records.
"""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.database import get_db
from ...core.error_handler import _build_response
from ...core.exceptions import BadRequestException
from ..deps import get_current_user
from ...schemas.qualification import (
    DraftRequest,
    QualificationCreate,
    QualificationUpdate,
    UploadTokenRequest,
    get_status_label,
)
from ...services import qualification_service

router = APIRouter(tags=["qualifications"])


# ──────────────────────────────────────────────────────────────────
# POST /qualification-files/upload-token
# ──────────────────────────────────────────────────────────────────
@router.post("/qualification-files/upload-token")
async def get_upload_token(
    data: UploadTokenRequest,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Generate a COS pre-signed upload URL for a qualification file.

    Validates file type (JPG/PNG/PDF) and size (max 10MB).
    Returns a fileId, uploadUrl, and expiry time.
    """
    user_id = int(current_user["sub"])
    result = await qualification_service.get_upload_token(
        user_id=user_id,
        file_name=data.fileName,
        content_type=data.fileType,
        file_size=data.fileSize,
    )
    return _build_response(0, "success", result)


# ──────────────────────────────────────────────────────────────────
# POST /qualifications (submit)
# ──────────────────────────────────────────────────────────────────
@router.post("/qualifications")
async def submit_qualification(
    data: QualificationCreate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Submit a qualification record for admin review.

    After submission, the qualification enters 'reviewing' status and
    triggers admin notification.
    """
    user_id = int(current_user["sub"])
    qualification = await qualification_service.submit_qualification(db, user_id, data)

    return _build_response(
        0,
        "success",
        {
            "qualificationId": str(qualification.id),
            "status": qualification.status.value if hasattr(qualification.status, "value") else str(qualification.status),
            "statusLabel": get_status_label(
                qualification.status.value if hasattr(qualification.status, "value") else str(qualification.status)
            ),
            "submittedAt": qualification.submitted_at.isoformat() if qualification.submitted_at else None,
        },
    )


# ──────────────────────────────────────────────────────────────────
# GET /qualifications/current
# ──────────────────────────────────────────────────────────────────
@router.get("/qualifications/current")
async def get_current_qualifications(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Get the current user's qualification records.

    Returns all qualifications (draft, reviewing, approved, rejected, etc.).
    """
    user_id = int(current_user["sub"])
    result = await qualification_service.get_current(db, user_id)
    return _build_response(0, "success", result)


# ──────────────────────────────────────────────────────────────────
# PUT /qualifications/{id}
# ──────────────────────────────────────────────────────────────────
@router.put("/qualifications/{qualification_id}")
async def update_qualification(
    qualification_id: int,
    data: QualificationUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Update and resubmit a qualification (only rejected or draft records).

    Uses optimistic locking via the version field.
    """
    user_id = int(current_user["sub"])
    qualification = await qualification_service.update_qualification(
        db, user_id, qualification_id, data
    )

    return _build_response(
        0,
        "success",
        {
            "qualificationId": str(qualification.id),
            "status": qualification.status.value if hasattr(qualification.status, "value") else str(qualification.status),
            "statusLabel": get_status_label(
                qualification.status.value if hasattr(qualification.status, "value") else str(qualification.status)
            ),
            "submittedAt": qualification.submitted_at.isoformat() if qualification.submitted_at else None,
        },
    )


# ──────────────────────────────────────────────────────────────────
# GET /qualifications/{id}/reviews
# ──────────────────────────────────────────────────────────────────
@router.get("/qualifications/{qualification_id}/reviews")
async def get_qualification_reviews(
    qualification_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Get the review history for a qualification record."""
    user_id = int(current_user["sub"])
    result = await qualification_service.get_reviews(db, user_id, qualification_id)
    return _build_response(0, "success", result)


# ──────────────────────────────────────────────────────────────────
# POST /qualifications/draft
# ──────────────────────────────────────────────────────────────────
@router.post("/qualifications/draft")
async def save_qualification_draft(
    data: DraftRequest,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """Save a qualification draft (does not trigger review).

    Creates or updates a draft. All fields are optional.
    """
    user_id = int(current_user["sub"])
    draft = await qualification_service.save_draft(db, user_id, data)

    return _build_response(
        0,
        "success",
        {
            "qualificationId": str(draft.id),
            "status": draft.status.value if hasattr(draft.status, "value") else str(draft.status),
            "statusLabel": get_status_label(
                draft.status.value if hasattr(draft.status, "value") else str(draft.status)
            ),
            "savedAt": draft.updated_at.isoformat() if draft.updated_at else None,
        },
    )

File: manageSystem/src/stores/qualifications.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import http from '@/api/http'
import { ElMessage } from 'element-plus'

export const useQualificationsStore = defineStore('qualifications', () => {
  // State
  const qualifications = ref([])
  const currentQualification = ref(null)
  const reviewList = ref([])
  const loading = ref(false)
  const reviews = ref([])

  // Status helpers
  function getStatusLabel(status) {
    const labels = {
      draft: '草稿',
      reviewing: '待审核',
      approved: '已通过',
      rejected: '已驳回',
      expiring: '即将过期',
      expired: '已过期',
    }
    return labels[status] || status
  }

  function getStatusType(status) {
    const types = {
      draft: 'info',
      reviewing: 'warning',
      approved: 'success',
      rejected: 'danger',
      expiring: 'warning',
      expired: 'info',
    }
    return types[status] || 'info'
  }

  // Actions
  async function fetchUploadToken(fileName, fileType, fileSize) {
    try {
      const res = await http.post('/qualification-files/upload-token', {
        fileName,
        fileType,
        fileSize,
      })
      return res.data.data
    } catch (e) {
      ElMessage.error(e.userMessage || '获取上传凭证失败')
      throw e
    }
  }

  async function fetchCurrent() {
    loading.value = true
    try {
      const res = await http.get('/qualifications/current')
      const data = res.data.data
      qualifications.value = data.items || []
      return data
    } catch (e) {
      ElMessage.error(e.userMessage || '获取资质信息失败')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function submitQualification(data) {
    loading.value = true
    try {
      const res = await http.post('/qualifications', data)
      const result = res.data.data
      ElMessage.success('资质提交成功，等待审核')
      await fetchCurrent()
      return result
    } catch (e) {
      if (e.response?.status === 409) {
        ElMessage.error('已有待审核或已通过的资质记录')
      } else {
        ElMessage.error(e.userMessage || '提交资质失败')
      }
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updateQualification(qualificationId, data) {
    loading.value = true
    try {
      const res = await http.put(`/qualifications/${qualificationId}`, data)
      const result = res.data.data
      ElMessage.success('资质已重新提交')
      await fetchCurrent()
      return result
    } catch (e) {
      if (e.response?.status === 409) {
        ElMessage.error('版本冲突：资质已被他人修改，请刷新后重试')
      } else {
        ElMessage.error(e.userMessage || '更新资质失败')
      }
      throw e
    } finally {
      loading.value = false
    }
  }

  async function saveDraft(data) {
    loading.value = true
    try {
      const res = await http.post('/qualifications/draft', data)
      const result = res.data.data
      ElMessage.success('草稿已保存')
      return result
    } catch (e) {
      if (e.response?.status === 409) {
        ElMessage.error('已有待审核或已通过的资质记录，无法保存草稿')
      } else {
        ElMessage.error(e.userMessage || '保存草稿失败')
      }
      throw e
    } finally {
      loading.value = false
    }
  }

  async function fetchReviews(qualificationId) {
    loading.value = true
    try {
      const res = await http.get(`/qualifications/${qualificationId}/reviews`)
      reviews.value = res.data.data.items || []
      return reviews.value
    } catch (e) {
      ElMessage.error(e.userMessage || '获取审核记录失败')
      throw e
    } finally {
      loading.value = false
    }
  }

  // Admin actions
  async function fetchReviewList(status) {
    loading.value = true
    try {
      const params = {}
      if (status) params.status = status
      const res = await http.get('/admin/qualifications', { params })
      reviewList.value = res.data.data.items || []
      return reviewList.value
    } catch (e) {
      ElMessage.error(e.userMessage || '获取审核列表失败')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function reviewQualification(qualificationId, action, comment) {
    loading.value = true
    try {
      const res = await http.post(`/admin/qualifications/${qualificationId}/review`, {
        action,
        comment: comment || '',
      })
      const result = res.data.data
      const actionLabel = action === 'approve' ? '通过' : '驳回'
      ElMessage.success(`资质审核${actionLabel}`)
      await fetchReviewList()
      return result
    } catch (e) {
      ElMessage.error(e.userMessage || '审核操作失败')
      throw e
    } finally {
      loading.value = false
    }
  }

  return {
    // State
    qualifications,
    currentQualification,
    reviewList,
    loading,
    reviews,
    // Status helpers
    getStatusLabel,
    getStatusType,
    // Actions
    fetchUploadToken,
    fetchCurrent,
    submitQualification,
    updateQualification,
    saveDraft,
    fetchReviews,
    fetchReviewList,
    reviewQualification,
  }
})
